File: programmers/python/level3/binary_search/입국심사.py
# https://school.programmers.co.kr/learn/courses/30/lessons/43238
def solution(n, times):
    times.sort()
    start = times[0]
    end = times[0] * n
    ans = -1
    
    while start <= end:
        mid = (start + end) // 2 ## mid 분까지 
        cover = 0
        for time in times:
            if time > mid:
                break
            cover += mid // time
        
        if cover >= n:
            ans = mid
            end = mid - 1
        else:
            start = mid + 1
    
    
    return ans
 

File: programmers/java/src/level3/binarysearch/입국심사.java
package level3.binarysearch;

import java.util.Arrays;

public class 입국심사 {

    public long solution(int n, int[] times) {
        Arrays.sort(times);

        long start = times[0];
        long end = (long) times[0] * n;
        long ans = -1;
        while( start <= end ) {
            long mid = (start + end) / 2;
            long covered = 0;
            for (int time : times) {
                if (time > mid) {
                    break;
                }
                covered += mid / time;
            }

            if (covered >= n) {
                ans = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return ans;
    }
}
